🌡️ feat: Periodic Health Check to prevent UI Inactivity Connection Errors

danny-avila opened the pull request:
## Summary

Closes #3579 

I introduced a periodic health check mechanism to prevent UI inactivity connection errors.

Here's a summary on the health check behavior:

1. Regular Interval Checks:
   - Performed every 10 minutes automatically.

2. Window Focus Checks:
   - Triggered when the window regains focus.
   - Only performs a new check if the last check was more than 10 minutes ago.

3. User Interaction Checks:
   - Triggered when the user types in the chat input.
   - Forces an immediate health check by invalidating the query if the last user interaction was more than 5 minutes ago.

4. No Caching:
   - Each health check fetches fresh data from the server.
   - No data is cached between checks.

5. Error Handling:
   - Failed checks are logged but not retried automatically.

### More Details

- Implemented `useHealthCheck` and `useInteractionHealthCheck` hooks in `client/src/data-provider/connection.ts` to perform health checks periodically and upon user interactions.
- Exported `useHealthCheck` and `useInteractionHealthCheck` in `client/src/data-provider/index.ts`.
- Integrated `useInteractionHealthCheck` into `useTextarea` hook to trigger health check on user interactions in `client/src/hooks/Input/useTextarea.ts`.
- Added `useHealthCheck` in `ChatRoute` component for periodic health checks in `client/src/routes/ChatRoute.tsx`.
- Defined health check API endpoint in `packages/data-provider/src/api-endpoints.ts`.
- Created `healthCheck` function in `data-service.ts` to call the health check API endpoint.
- Added `health` key to `QueryKeys` enum in `packages/data-provider/src/keys.ts`.


### Other changes

- Programmatically scroll to end when a chat request is initiated (and messages have rendered)
    - prevents issues where automatic scrolling fails to initiate if the user starts a request while the latest messages are not in view

### Testing

To test these changes, I recommend:

1. **Unit Tests**: Write unit tests for the `useHealthCheck` and `useInteractionHealthCheck` hooks to ensure they perform health checks correctly.
2. **Integration Tests**: Test the integration of these hooks in the `useTextarea` and `ChatRoute` components to verify that health checks are triggered appropriately.
3. **Manual Testing**: Ensure the application remains connected without UI inactivity connection errors by simulating periods of inactivity and user interactions.

### Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes

danny-avila committed:
🌡️ feat: Periodic Health Check to prevent UI Inactivity Connection Errors

danny-avila committed:
feat: Add refetchOnWindowFocus option for health check

danny-avila committed:
feat: programmatically scroll to end when a chat request is initiated (and messages have rendered)

GerkinDev commented:
@danny-avila IMHO this means that the underlying types are not accurate, and some nullish value are not typed as such. Could you please give me an example of a file demonstrating what troubles you, so that I can play around and see what I can do for it ?

I don't think that allowing the exact opposite issue like the bug I fixed is a satisfying solution either.

danny-avila commented:
> @danny-avila IMHO this means that the underlying types are not accurate, and some nullish value are not typed as such. Could you please give me an example of a file demonstrating what troubles you, so that I can play around and see what I can do for it ?
> 
> I don't think that allowing the exact opposite issue like the bug I fixed is a satisfying solution either.

Sure there might be issues, but I'm making a lot of changes at the moment in prep for a release, so I don't want to accidentally change anything I don't intend to change. I only realized this was happening during the work for this PR.

There was one situation I don't remember but it caused an `undefined` error somewhere along development. In that case, I did not "allow" it and fixed it according to the lint rule but I am wary of creating similar situations until after release.

GerkinDev commented:
OK, don't hesitate to ping me if you see it again so that I can give you a hand or after the release so that we can dig out if it's possible to re-enable them
